src/db.rs

```diff
@@ -2,154 +2,218 @@
+            // entry.
+            info!("{} is already a member of {}",
+                  record.client.clone(), record.public_ip.clone());
+            let _: () = try!(
+                cmd("SET").arg(key.clone())
+                          .arg(record.message.clone())
+                          .query(&self.connection)
+            );
+        }
 
-    #[cfg(test)]
-    pub fn clear(&self) -> rusqlite::Result<()> {
-        self.connection.execute_batch(
-            "DELETE FROM boxes;
-             VACUUM;"
-        )
+        // And set the TTL of the message.
+        // The entry in the list of clients for this public IP will be
+        // cleaned up during the .get call. It does no harm to keep it
+        // around until that point.
+        let _: () = try!(
+            cmd("EXPIRE").arg(key.clone())
+                         .arg(115) // 115 seconds.
+                         .query(&self.connection)
+        );
+
+        Ok(())
     }
 
-    // Looks for records for a given constraint.
-    pub fn find(&self, filter: FindFilter) -> rusqlite::Result<Vec<Record>> {
-        let mut stmt: rusqlite::Statement;
-
-        let rows = match filter {
-            FindFilter::PublicIp(public_ip) => {
-                stmt = try!(
-                    self.connection.prepare("SELECT * FROM boxes WHERE public_ip=$1")
-                );
-                try!(stmt.query(&[&escape(&public_ip)]))
-            },
-            FindFilter::PublicIpAndClient(public_ip, client_id) => {
-                stmt = try!(
-                    self.connection.prepare("SELECT * FROM boxes WHERE (public_ip=$1 and client=$2)")
+    ///
+    /// Get the registration entries for a given public IP.
+    ///
+    pub fn get(&self, public_ip: String) -> RedisResult<Vec<Record>> {
+        let _: () = try!(
+            cmd("WATCH").arg(public_ip.clone())
+                        .query(&self.connection)
+        );
+
+        // Get the clients for the given public IP.
+        let members: Vec<String> = try!(
+            cmd("SMEMBERS").arg(public_ip.clone())
+                           .query(&self.connection)
+        );
+
+        info!("Members of {}: {:?}", public_ip.clone(), members.clone());
+
+        let mut result = Vec::new();
+
+        // For each client we get the associated message.
+        for member in members {
+            let key = format!("{}:{}", public_ip.clone(), member);
+            info!("Key {}", key.clone());
+            let message = match cmd("GET").arg(key.clone())
+                                           .query(&self.connection) {
+                Ok(message) => message,
+                Err(_) => "".to_string()
+            };
+
+            info!("Message for {}: {}", key.clone(), message.clone());
+
+            // Remove the client id from the list of clients of this public
+            // IP that has no associated message.
+            if message.is_empty() {
```

Comment:
That would be more idiomatic to make message be an Option<String> and to set it to None in the Err(_) case.